core/server/kustomization_test.go

```diff
@@ -104,7 +106,7 @@ func TestListKustomizationPagination(t *testing.T) {
 		},
 	})
 	g.Expect(err).ToNot(HaveOccurred())
-	g.Expect(res.Kustomizations).To(HaveLen(2))
+	g.Expect(res.Kustomizations).To(HaveLen(existingKust + 2))
```

Comment:
@luizbafilho The refactor you did on pagination looks really cool. It's a lot cleaner and easier to read. 

Just a small question. Should this assertion be `g.Expect(res.Kustomizations).To(HaveLen(existingKust + 1))`instead? I mean, in the previous call to `ListKustomization` you have PageSize = 1.

Reply:
that's the caveat with this whole multicluster, multinamespace querying, we need to multiplicate things, so it's accounting for 1 kustomization for each of the namespaces, summing 2.

Reply:
Hmm. I see, So the `PageSize` parameter means, in this case, "bring me 1 kustomization from each namespace" ?.  To be honest that is different from the original plan. My understanding is that the pageSize is the length of the page  I'm requesting (as long as there are enough resources of course).

Reply:
That's the natural paging system, but that's not doable when having multiple parallel requests going for each cluster and each namespace, it would require them to be sequential which is not viable.

We can think of our paging just a way to avoid user waiting too much before getting anything and our goal just to get a bunch more data when the user presses the `next` button.

Reply:
Hmm, that's interesting. 

Ok, just recapitulating a bit to see if I understood this, properly. 

The problem we are trying to solve is `user waiting too much before waiting too much`, I agree 100% ✅. Now, the approach you are following to solve that, is by having `multiple parallel requests`, where each request would be paginated(limited to a specific size), correct?. 

If so, then I must say that I really like this approach because it is aligned to what we are trying to solve. 

However, I see a few implications to consider.

1) Sorting. Let's pretend limit is 2 (2 kustomizations from each namespace)

If the first would look like this:
```
Namespace KustomizationName
      ns1             a1
      ns1             a2
      ns2            b1
      ns2            b2
```

and second page like this:
```
Namespace KustomizationName
      ns1             z1
      ns1             z2
      ns2            z1
      ns2            z2
```

I wonder if it would be a good user experience to still receive kustomizations from namespace `ns1` in the second page. I personally, after seeing page 1 elements, would think that there should be no more kustomizations for namespace `ns1` in the next page. In the previous approach you will get the pages this way:

Page  1
```
Namespace KustomizationName
      ns1             a1
      ns1             a2
      ns1             z1
      ns1             z2
```

Page 2
```
Namespace KustomizationName
      ns2            z1
      ns2            z2
      ns2            b1
      ns2            b2
```

Here, everything is sorted first by namespace and then by the name of the kustomization, ensuring that every page is alphabetically ordered with each other. Which can solve the issue exposed in the previous statement.

2) Page size. Somehow, in the UI. We would need to tell the user, how many kustomizations are going to be available in every page. In the future this could be even configurable. Following the approach on this PR, we can't really ensure that amount of items per page, even if the actual total amount of kustomizations(across namespaces) can be divided into many pages. Unless we explicitly say that this limit is per namespace, which could also lead to have different number of kustomizations in each page, as namespaces may not have the exact amount requested in the PageSize parameter.

3) Complexity. In terms of pagination, I think having simultaneous requests add a bit of unnecessary complexity. If we want the sort to use namespace name and kustomization name, then we can simply just make 1 request and that would be it (assuming enough kustomization to fullfil pageSize). And continue from there in each consecutive request.

Proposal

Now my approach has its disadvantages as well. If I want to query 100 kustomizations across 10 namespaces that have 10 kustomizations each. That would result, in a very slow request as I will be making 10 request to the k8s api sequentially.  

So, considering all this problems from both approaches. I strongly believe we can combine both approaches and have a more complete solution. The key is going to rely on knowing the total amount of kustomizations in each namespace. If we have this information available then, considering the parameters in the page request we can calculate how many requests I will need and to which namepaces pointing it to. 

For instance: 

Data
```
Namespace   Kustomization Name
       .                          
       .
       .
       ns1                      a1
       ns1                      b1
       ns1                      c1
       ns2                      a2
       ns2                      b2
       ns2                      c2
       ns3                      a3
       ns3                      b3
       ns3                      c3
       ns4                      a4
       ns4                      b4
       ns5                      a5
       ns5                      b5
       ns5                      c5
```

Request parameters
```
pageSize: 6
pageToken: XYZ
```

Now, let's say the pageToken received say that we need to continue from namespace namespace `ns3` and just after kustomization `b3`. As the parameter pageSize is `6` it means we need to query 6 kustomizations starting from there. So, by knowing the total amount of kustomizations in each namespace we can calculate that we will need to make 3 parallel request:

1.- Request 1 kustomizations from namespace `ns2` --> Will return `c3`
2.- Request 2 kustomizations from namespace `ns3` --> Will return `a4` and `b4`
3.- Request 3 kustomization from namespace `ns5` --> Will return `a5`, `b5` and `c5`

Now, the 3 previous implications are personal.  I would like to ask @sympatheticmoose his point of view regarding this matter, as I could be wrong on these assumptions. I'm also open to discuss this further in a meeting with the  rest of the team and product. 

I must say, I also have the plan to talk with product next week, about some other issues I saw on my spike regarding pagination. I'll leverage that talk and I'll bring up your approach as well. I just would like to involve everyone on this and conclude something before anyone can start writing more code about it. This way we can safe people's effort and time.

Reply:
Right, those are the kind of issues I was referring to. Kube API actually supports some counting, sorting and filter functionality, but they come with some important limitations. That's why I have a meeting setup for next Wednesday with Product, to express these restrictions. 

Now, just to make sure I'm understanding properly. When you say it would be ideal to rely on the UI for the sorting, counting, and filtering, what do you actually mean? Are you saying that the UI will have all the data, maybe in memory(e.g. all kustomizations) and then will do the sorting/counting/filtering in there?.

> The only way to make that happen would have all data from the cluster synced to a central queriable data source, but I have the impression this is out of the table due to security constraints.

Interesting. But I'm confused. What are the security constraints you are referring to?. Are you, perhaps, referring to the location of that data source? Would the cluster itself be insecure ? Would it be safe to use RBAC to secure that data? (k8s student here 😋).

> But if you find I way to do it, we can discuss it later and implement it in follow-up PR.
Actually, my plan is to first ask Product what we really want regarding this matter and then probably have a brainstorm meeting with the team, so everyone is involved and we try to come up with some ideas on how to implement it.

Reply:
Do you mind dropping documents on how you sort a query or only get the number of items of a particular list without making the list call?

I didn't say it would be _ideal_ to really be on the ui for sorting, counting, and filtering.  I've said that since the Kube API
> does not provide any functionality regarding counting, sorting, or filtering, so that must be done in the UI only

To accomplish that kind of filtering or searching, we would need to centralize all the data from all the clusters in one single place (ie. elasticsearch) and use that as a data source instead of querying the clusters on the fly. Having all that data in one single place is potentially dangerous and hard to manage properly so users probably want to avoid that, we would also need to build our RBAC to guarantee the users only access the data they can.

Reply:
On the counting is that there is no specific call to count things as we would in SQL (`SELECT COUNT(*) FROM table_name;`) and we would that number even before making the first call to distribute the `limit` across the clusters/namespaces calls.

And as the paragraph states, if we use any label or field selector the `remainingItemCount` won't be there, so we can't rely on it.